# Assignment 6: Using Pandas to load and manipulate spreadsheets

## Instructions

1. Fill in the code where indicated to complete the assignment. Feel free to add more cells than provided to try things out, and check what your variables look like, to see if you got the right results. Just make sure you have done what is asked for that question and that you don't delete any code we provide you with.

2. Rename this `last_first_assignment6.ipynb` file replacing `last` and `first` with your name. Make sure all the outputs are there (run each cell) and then export as `last_first_assignment6.py` as well.

3. Submit both these files to Quercus.

## Questions and Support
1. Please ask class content questions on the class GitHub page: 
https://github.com/BME1478H/Fall2020class/ Expect a processing time of 1d for us to respond. As such, we cannot guarantee we can address your questions and work through all the troubleshooting in time if you ask them the night before or the day of the assignment deadline. Please plan accordingly. 

## Part 1: Manipulating and subsetting dataframes (4 marks)

Here we are setting up our notebook to import Pandas and load the dataset used in the lecture material. **Run this cell to get started.**

In [2]:
import pandas as pd
import numpy as np
url = 'https://raw.githubusercontent.com/BME1478H/Winter2022class/master/data/world-data-gapminder.csv'
world_data = pd.read_csv(url)

a) Subset/select the following columns from the `world_data` dataframe:
- `country`
- `life_expectancy`
- `co2_per_capita`

Store the result in a variable called `world_subset`. 

Write the code in the cell below. (0.5 marks)

In [3]:
world_data.loc[:, ['country', 'life_expectancy', 'co2_per_capita']] #obtain columns for country, life_expectancy, co2_per_capita and all rows
world_subset = world_data.loc[:, ['country', 'life_expectancy', 'co2_per_capita']] #store results in world_subset
print(world_subset) #print world_subset

           country  life_expectancy  co2_per_capita
0      Afghanistan             28.2             NaN
1      Afghanistan             28.2             NaN
2      Afghanistan             28.2             NaN
3      Afghanistan             28.2             NaN
4      Afghanistan             28.2             NaN
...            ...              ...             ...
39196     Zimbabwe             57.0            0.78
39197     Zimbabwe             58.3             NaN
39198     Zimbabwe             59.3             NaN
39199     Zimbabwe             59.8             NaN
39200     Zimbabwe             60.2             NaN

[39201 rows x 3 columns]


b) Slice `world_data` such that we get the first 12 rows, and the last three columns using `dataframe.iloc[]` notation. Store the result in a variable called `world_slice`. 

___
**Hint:** Slicing using indices takes the indices starting from the "start" slice value, UP TO but NOT INCLUDING the "end" slice value. For example:

```
world_data.iloc[5:10] 
```
This will print  5, 6, 7, 8, and 9 (but not 10).

**Hint:** This is one way to slice last `4` items:
```
lst = [0,1,2,3,4,5,6]
lst[-4:] 

Output: [3,4,5,6]
```
The -4 means the 4th last item, and leaving it blank after the : means go to the end!
___

Write the code in the cell below. (0.5 marks)

In [4]:
world_data.iloc[0:13]  #first 12 rows
lst = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14] #number of columns
lst[-3:] #last 3 columns
world_data.iloc[0:13, -3:] #ouput first 12 rows and last three columns

co2_per_capita  years_in_school_men  years_in_school_women
0              NaN                  NaN                    NaN
1              NaN                  NaN                    NaN
2              NaN                  NaN                    NaN
3              NaN                  NaN                    NaN
4              NaN                  NaN                    NaN
5              NaN                  NaN                    NaN
6              NaN                  NaN                    NaN
7              NaN                  NaN                    NaN
8              NaN                  NaN                    NaN
9              NaN                  NaN                    NaN
10             NaN                  NaN                    NaN
11             NaN                  NaN                    NaN
12             NaN                  NaN                    NaN

c) In what situations would you use `world_data.iloc[]` vs. `world_data.loc[]`? What is the difference in syntax?

Write out your response in the following variable inside the quotes `'''` we defined for you (triple quotes  allow us to write on more than one line).
1-2 sentences max (1 mark)

In [5]:
answer = '''world_data.iloc[] is used to select rows/columns by integer postion (eg; world_data.iloc[0,1]) - it uses integer indexes. 
world_data.loc[] is used to select rows/columns by label position (eg; world_data.loc['country', 'year']) - it uses labels.'''
print(answer)

world_data.iloc[] is used to select rows/columns by integer postion (eg; world_data.iloc[0,1]) - it uses integer indexes. 
world_data.loc[] is used to select rows/columns by label position (eg; world_data.loc['country', 'year']) - it uses labels.


d) From `world_data`, select the row labelled 100, from the column label `year`. Use the correct notation to subset and store the value inside the variable `year_100_value`.

Write the code in the cell below. (1 mark). 

In [6]:
world_data.loc[100, 'year'] #row 100 and label year
year_100_value = world_data.loc[100, 'year'] #store result in variable year_100_value
print(year_100_value) #print year_100_value

1900


e) From `world_data`, select rows labeled 5 to 10, and columns 'country' to 'income_group' using slicing.

Store in a variable called `world_slice_2`

Write the code in the cell below. (1 mark).

In [7]:
world_data.loc[5:10, 'country': 'income_group']  #row 5 to 10 with column country to income_group
world_slice_2 = world_data.loc[5:10, 'country': 'income_group']  #store result in variable world_slice_2
print(world_slice_2) #print world_slice_2

        country  year  population region     sub_region income_group
5   Afghanistan  1805     3280000   Asia  Southern Asia          Low
6   Afghanistan  1806     3280000   Asia  Southern Asia          Low
7   Afghanistan  1807     3280000   Asia  Southern Asia          Low
8   Afghanistan  1808     3280000   Asia  Southern Asia          Low
9   Afghanistan  1809     3280000   Asia  Southern Asia          Low
10  Afghanistan  1810     3280000   Asia  Southern Asia          Low


# Part 2: Filtering observations and creating new columns (3 marks)

In the material, we saw we could filter the data based on the criteria we  define. For example:
```
above_83_bool = world_data['life_expectancy'] > 83
```
This would result in a boolean array stored inside `above_83_bool` where every row shows either `True` or `False` depending whether the `life_expectancy` value in that row was greater than 83. We can use these boolean arrays to further subset our data.
___

a) Create a boolean array from `world_data` where we are checking whether the `year` column is equal to `2014`, and store in a variable called `year_2014`.

Write the code in the cell below. (0.5 marks).

In [8]:
world_data['year'] == 2014   #get data where year equals 2014
year_2014 = world_data['year'] == 2014  #store in variable year_2014
print(year_2014) #print boolean array of year_2014

#checking if number of rows or True is correct
year_2014_bool = world_data['year'] == 2014   #boolean variable to get year_2014 data where year equals 2014
year_2014_bool.sum() #sum of number of rows (True values)

0        False
1        False
2        False
3        False
4        False
         ...  
39196     True
39197    False
39198    False
39199    False
39200    False
Name: year, Length: 39201, dtype: bool


179

b) Use the boolean array `year_2014` you just created to subset the `world_data` dataframe and store in a variable `year_2014_subset`. 

The resulting array should be of size (179 rows, 14 columns). 
___
**Hint:** We can subset a dataframe with our boolean expression using similar notation to row indexing of dataframes. From lecture:
```
world_data[world_data['life_expectancy'] > 83]
```
___

Write the code in the cell below. (0.5 marks).

In [9]:
year_2014_subset = world_data['year'] == 2014  #store result of subset in variable year_2014_subset
print(year_2014_subset)  #print boolean array of year_2014_subset

#checking if number of rows or True is correct
year_2014_subset_bool = world_data['year'] == 2014  #boolean variable to get year_2014_subset data where year equals 2014
year_2014_subset_bool.sum() #sum of number of rows (True values)

0        False
1        False
2        False
3        False
4        False
         ...  
39196     True
39197    False
39198    False
39199    False
39200    False
Name: year, Length: 39201, dtype: bool


179

c) Use the boolean arrays `criterion_1` and `criterion_2` we provide below to subset `world_data` such that we get rows that match BOTH these criteria.

Store in a variable called `both_criterion_subset`.
___
**Hint:** When we first learned about booleans, we learned we can use `and` and `or` operators to evaluate whether two booleans were both true (and) and either are True (or). 
```
x = (1 > 0) # True, 1 is greater than 0
y = (5 > 10) # False, 5 is not greater than 10
z = (1 == 1) # True, 1 is equal to 1

x and z # this evaluates to True, x and z BOTH True using "and"
x and y # False because only one value, x is True, Y is False using "and"
y or z # True because only one value, x, needs to be True using "or" 
x or z # True, both are True, but only one needs to be True anyway
```
But for Dataframes we need to use the `&` operator instead of `and` to match all criteria or the `|` operator instead of `or` to match _any_ of our criteria.
```
df[(df['column0'] > 1) & (df['column4'] == 'dog')]
```
This subsets our imaginary `df` dataframe by rows where `column0` is greater than 1 AND where `column4` is equal to `dog`. 
___

Write your code below our code in the same cell (0.5 marks).

In [10]:
criterion_1 = world_data['income'] > 50000
criterion_2 = world_data['life_expectancy'] > 80

# write your code below
both_criterion_subset = world_data[criterion_1 & criterion_2]  #store result of rows that match both criteria in varilable both_criterion_subset
print(both_criterion_subset) #print both_criterion_subset 

           country  year  population  region       sub_region income_group   
16640      Ireland  2015     4700000  Europe  Northern Europe         High  \
16641      Ireland  2016     4730000  Europe  Northern Europe         High   
16642      Ireland  2017     4760000  Europe  Northern Europe         High   
16643      Ireland  2018     4800000  Europe  Northern Europe         High   
20573   Luxembourg  2006      466000  Europe   Western Europe         High   
...            ...   ...         ...     ...              ...          ...   
33721  Switzerland  2014     8230000  Europe   Western Europe         High   
33722  Switzerland  2015     8320000  Europe   Western Europe         High   
33723  Switzerland  2016     8400000  Europe   Western Europe         High   
33724  Switzerland  2017     8480000  Europe   Western Europe         High   
33725  Switzerland  2018     8540000  Europe   Western Europe         High   

       life_expectancy  income  children_per_woman  child_morta

d) Let's say we wanted to extract some measures of health (life expectancy and child mortality) in high income countries in a specific year. We would start by subsetting the data. 

For this question, subset `world_data` by the following boolean arrays so we match ALL the criteria, and also selecting the given column labels (remember, when we subset rows AND columns using labels, there is a specific notation to use). 

Store the result in a variable called `high_income_health`. 

Write the code below our premade code in the same cell (0.5 marks). 

In [11]:
criterion_1 = world_data['income'] > 50000
criterion_2 = world_data['year'] == 2013
column_subset = ['country', 'life_expectancy', 'child_mortality']

# write your code here
world_data.loc[criterion_1 & criterion_2, column_subset]  #store result of both criterias and column_subset
high_income_health = world_data.loc[criterion_1 & criterion_2, column_subset] #store result in variable high_income_health
print(high_income_health) #print  high_income_health

                    country  life_expectancy  child_mortality
18609                Kuwait             79.7              9.5
20580            Luxembourg             81.9              2.6
26055                Norway             81.7              2.9
28464                 Qatar             80.5              8.7
30654             Singapore             83.2              2.7
33720           Switzerland             82.7              4.3
36786  United Arab Emirates             76.4              8.0
37437         United States             78.9              6.9
37438         United States             78.9              6.9


e) Subset `world_data` such that only rows where the `pop_density` is greater than `1000` OR where the `co2_per_capita` value is greater than `20`, but also only selecting the columns `['country', 'year', 'pop_density', 'co2_per_capita']`. The resulting shape should be (447,4). 

Don't forget to wrap each boolean expression with `(` `)`.

Store in variable called `density_or_co2`.

Write the code below (0.75 marks).

In [12]:
criterion_1 = world_data['pop_density'] > 1000  #criteria 1
criterion_2 = world_data['co2_per_capita'] > 20  #criteria 2
column_subset = ['country', 'year', 'pop_density', 'co2_per_capita']  #selected columns of country, year, pop_density, co2_per_capita

density_or_co2 = world_data.loc[criterion_1 | criterion_2, column_subset] #store result of both criterias using or and column_subset in variable density_or_co2
print(density_or_co2)  #print density_or_co2

             country  year  pop_density  co2_per_capita
2361         Bahamas  1971         17.4            38.7
2362         Bahamas  1972         17.8            36.5
2363         Bahamas  1973         18.1            43.4
2364         Bahamas  1974         18.5            39.9
2365         Bahamas  1975         18.9            43.7
...              ...   ...          ...             ...
37370  United States  1979         24.9            21.5
37371  United States  1980         25.1            20.6
37372  United States  1980         25.1            20.6
37411  United States  2000         30.8            20.2
37412  United States  2000         30.8            20.2

[447 rows x 4 columns]


f) Create a new column in `world_data` called `pop_millions` that divides the `population` column by `1000000` to get a more readable value for the population in terms of millions of people.
___
**Hint**: This is an example of creating a new column called `sizes_sum` where we want to add the values in two columns:
```
df['sizes_sum'] = df['size_1'] + df['size_2']
```
___

Write your code below. (0.25 marks). 

In [13]:
world_data['pop_millions'] = world_data['population'] / 1000000 #create new column with label pop_millions, take population and divide by 1000000
world_data[['population', 'pop_millions']] #have values in two columns

population  pop_millions
0         3280000          3.28
1         3280000          3.28
2         3280000          3.28
3         3280000          3.28
4         3280000          3.28
...           ...           ...
39196    15400000         15.40
39197    15800000         15.80
39198    16200000         16.20
39199    16500000         16.50
39200    16900000         16.90

[39201 rows x 2 columns]

# Part 3: Split-apply-combine using groupby (1 mark)

(summary from  lecture)

pandas facilitates the split-apply-combine workflow through the use of `groupby()` to split data, and summary/aggregation functions such as `mean()`, which collapses each group into a single-row summary of that group.

For example:

`world_data.groupby('region')['population'].sum()`

This groups `world_data` by the `region` column (i.e. 'Asia', 'Europe', 'Africa', 'Americas', 'Oceania') and calculates the `sum` of the `population` column for each of these `regions`.

___
a) Subset `world_data` to include only rows where the `region` is "Europe" and store in a variable `europe_subset`. (0.25 marks)

b) Group `europe_subset` by `sub_region` and find the mean `pop_density` for each of the European sub regions, store in `european_pop_density`. (0.5 marks)

c) Normalize (i.e. divide all values) in `european_pop_density` using the value from the subregion `Northern Europe`. As a result, `Northern Europe` will have a value of 1 (all values relative to that of `Northern Europe`.  (0.25 marks)

Write the code all the questions below.

In [14]:
#a)
world_data[world_data['region'] == "Europe"]  #set region equals to Europe
europe_subset = world_data[world_data['region'] == "Europe"]  #store result in variable europe_subset
print(europe_subset) #print europe_subset

#b)
european_pop_density = europe_subset.groupby('sub_region')['pop_density'].mean() #group europe_subset by sub_region, find mean of pop_density and store in variable european_pop_density
print(european_pop_density) #print european_pop_density

#c)
northern_europe_value = european_pop_density['Northern Europe']  #get pop_density value of northern europe
print(northern_europe_value) #print pop_density of northern europe
normalize_value = european_pop_density/ northern_europe_value #take northern europe pop_density value, divide all values and store it in variable normalize_value
print(normalize_value) #print normalize_value

              country  year  population  region       sub_region   
219           Albania  1800      410000  Europe  Southern Europe  \
220           Albania  1801      412000  Europe  Southern Europe   
221           Albania  1802      413000  Europe  Southern Europe   
222           Albania  1803      414000  Europe  Southern Europe   
223           Albania  1804      416000  Europe  Southern Europe   
...               ...   ...         ...     ...              ...   
37006  United Kingdom  2014    65000000  Europe  Northern Europe   
37007  United Kingdom  2015    65400000  Europe  Northern Europe   
37008  United Kingdom  2016    65800000  Europe  Northern Europe   
37009  United Kingdom  2017    66200000  Europe  Northern Europe   
37010  United Kingdom  2018    66600000  Europe  Northern Europe   

       income_group  life_expectancy  income  children_per_woman   
219    Upper middle             35.4     667                4.60  \
220    Upper middle             35.4     667   